lib/puppet/util/selinux.rb

```diff
@@ -47,8 +47,25 @@ def get_selinux_default_context(file)
     rescue Errno::EACCES, Errno::ENOENT
       mode = 0
     end
-
-    retval = Selinux.matchpathcon(file, mode)
+    # matchpathcon does only act on the passed string and not resolve the real path
+    # problematic if puppet manages /etc/init.d/script on RHEL which symlinks
+    # /etc/init.d to /etc/rc.d/init.d. matchpathcon will return the wrong label.
+    #
+    # This mimics restorecon behaviour.
+    #
+    # Not all files requested exist. Fall back to the requested file string.
+    begin
+      if !File.symlink?(file) && File.directory?(file)
+        file_realpath = File.realpath(file)
+      else
+        dirname_realpath = File.realpath(File.dirname(file))
+        file_basename = File.basename(file)
+        file_realpath = dirname_realpath + '/' + file_basename
+      end
+    rescue
```

Comment:
Is there an exception that can be caught here rather than just a general rescue?  This may hide legitimate bugs and should be avoided; always rescue the most specific exception one is able to, thus avoiding unintentional exception swallowing.